environmentd: add `storage_usage_retention_threshold`

necaris opened the pull request:
Add a `storage_usage_retention_threshold` configuration parameter which takes a Unix epoch timestamp. If this is set, records collected before the cutoff time are discarded when being read in from the Stash at boot.

Note that a sufficiently robust solution would ideally also handle #17548; this is not that solution (yet), although I would welcome feedback that evolves it that way.

### Motivation

  * This PR adds a known-desirable feature: Closes https://github.com/MaterializeInc/cloud/issues/3716

### Tips for reviewer



### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

necaris commented:
Note that this could be (should be?) expressed as an offset rather than a timestamp (e.g. X days old); this PR is up as a draft to ensure that other design discussions are hashed out, if necessary.

jkosh44 commented:
> Note that this could be (should be?) expressed as an offset rather than a timestamp (e.g. X days old); this PR is up as a draft to ensure that other design discussions are hashed out, if necessary.

There are a couple of places where we use this type, `EpochMillis`, to semantically represent a duration of time, but I do think using the `Duration` type in this case (and most of the other cases) would be better. 

Also we might want to leave a comment somewhere that the table is only actually truncated during a restart and not while running.

I also think it might be good to add a test. Here are some existing tests for storage usage that you can base it off of:
https://github.com/MaterializeInc/materialize/blob/b336a3255849f3e0fdfabd71eec3658dd06a3b6d/src/environmentd/tests/server.rs#L433-L670

necaris commented:
Thank you @jkosh44 for the review! Are there any existing tests for the bootstrap process that I could learn from, as well?

jkosh44 commented:
> Thank you @jkosh44 for the review! Are there any existing tests for the bootstrap process that I could learn from, as well?

I'm not sure of any tests that test specifically the bootstrap process. However, every time you create a new `Server` struct in the tests in `server.rs` via `let server = util::start_server(config).unwrap(); `, the bootstrap process will be run. I was thinking that you could have a test that had the following structure.

1. Set the storage usage interval and retention period to fairly small numbers.
2. Start up the server and wait for the first storage usage.
3. Stop the server and wait for the retention period to pass.
4. Start up another server and assert that the storage usage you saw previously is gone.

necaris commented:
Ah, perfect, thanks @jkosh44 !

mjibson commented:
Do you also want this to filter out old audit log events? If so, it could be renamed to represent it covers multiple collections.

I support having a test, but we can do better than waiting. We have carefully plummed a timestamp provider all through the codebase exactly for this reason. You can pass a custom object that spits out timestamps so you can make your server (`with_now` on the test util server config), shut it down, manually advance the timestamp, immediately start it up again and query the tables that should have been filtered.

mjibson commented:
```
    let now = Arc::new(Mutex::new(1));
    let now_fn = {
        let timestamp = Arc::clone(&now);
        NowFn::from(move || *timestamp.lock().unwrap())
    };

    let config = util::Config::default()
        .with_now(now_fn);
    let server = util::start_server(config).unwrap();

    *now.lock().unwrap() = 2;
```

Is an example of how to do this. It should work but...we don't have any current tests that need this, so it's possible something needs the timestamp to move forward actively (coord group commit maybe), unsure.

necaris commented:
Thanks @mjibson !

necaris commented:
So I have a (failing) test -- I initially tried this using @mjibson 's suggested method, and couldn't get it past a hang, so it now just waits. I'll keep working on debugging this failing test but if there's anything obvious I'm missing I'd love comments!

mjibson commented on src/environmentd/tests/server.rs:
```diff
@@ -669,6 +671,80 @@ fn test_storage_usage_collection_interval_timestamps() {
+
+        let subsequent_server_usage_records = Retry::default()
+            .max_duration(Duration::from_secs(5))
+            .retry(|_| {
```
Does this need to be retried?

mjibson commented on src/adapter/src/catalog.rs:
```diff
@@ -2714,7 +2714,17 @@ impl Catalog {
+                Some(period) => {
+                    let period_millis = period.as_millis();
+                    let moment = (config.now)();
+                    let VersionedStorageUsage::V1(e) = event;
```
`event.timestamp()` is a forward-compatible way to do this this is guaranteed to exist for any future version.

benesch commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -450,6 +450,11 @@ pub struct Args {
+    /// retention period is only evaluated at server start time, so rebooting
+    /// the server is required to discard old records
+    #[clap(long, env = "STORAGE_USAGE_RETENTION_PERIOD", parse(try_from_str = humantime::parse_duration))]
+    storage_usage_retention_period_timestamp: Option<Duration>,
```
Also the `_timestamp` seems redundant? I think just "storage-usage-retention-period" is clearer? And if we want it to be general purpose, perhaps "system-metadata-retention-period" or something like that?

necaris replied:
My concern with using it for the audit log is that currently, to bill for a cluster we need to retain a log of when it was created, no matter how long ago. We could change the billing query to `COALESCE()` to the beginning of the retention period if no creation record is found, and allow this to be used for both -- @benesch do you have a preference? I'm happy to take that tweak on in the billing code.

necaris commented:
@mjibson @benesch thank you both for the thorough review! I've addressed feedback and got the test passing, so I'd say this is ready for another look.

The question of whether to use this for audit log / other metadata is still open -- I wouldn't want to merge a change to the audit log without tweaking the query in the usage exporter to deal with the lack of a creation record (if it's beyond the retention period) but would be happy to do that as a prerequisite to releasing this.

mjibson commented on src/adapter/src/catalog.rs:
```diff
@@ -2714,7 +2714,16 @@ impl Catalog {
+        // TODO(necaris): remove this `as`. 584_542_046 years should be enough retention for anyone
+        let retention_period_ms = config
+            .storage_usage_retention_period
+            .map(|p| p.as_millis() as u64)
```
Keep this as a u128 which removes the `as` (we have banned `as`). Then use `u128::from(e.timestamp())` in the comparison below.

benesch commented:
Maybe I'm missing something, but this PR seems to be lacking the key part where the expired storage usage records are actually _removed_ from the stash. As far as I can tell, the expired records are simply not emitted into `mz_storage_usage_by_shard` table, but they remain in the stash.

mjibson commented:
> Maybe I'm missing something, but this PR seems to be lacking the key part where the expired storage usage records are actually _removed_ from the stash. As far as I can tell, the expired records are simply not emitted into `mz_storage_usage_by_shard` table, but they remain in the stash.

This is discussed in the commit message, so I had assumed the solution in this PR was aware of that:

> Note that a sufficiently robust solution would ideally also handle https://github.com/MaterializeInc/materialize/issues/17548; this is not that solution (yet), although I would welcome feedback that evolves it that way.

----

Is it safe to delete things from the stash instead? Are we sure the billing code won't have a problem with that? This PR has the benefit that we could still go get the data if we really needed to.

benesch commented:
> > Maybe I'm missing something, but this PR seems to be lacking the key part where the expired storage usage records are actually _removed_ from the stash. As far as I can tell, the expired records are simply not emitted into `mz_storage_usage_by_shard` table, but they remain in the stash.
> 
> This is discussed in the commit message, so I had assumed the solution in this PR was aware of that:
> 
> > Note that a sufficiently robust solution would ideally also handle #17548; this is not that solution (yet), although I would welcome feedback that evolves it that way.

I saw that paragraph, but it seems to be about a similar issue in unrelated tables (source and sink status histories), and not about the problem of leaving the expired data on disk. 

> 
> Is it safe to delete things from the stash instead? Are we sure the billing code won't have a problem with that? This PR has the benefit that we could still go get the data if we really needed to.

The whole point was to avoid accumulating data forever! Not because users wouldn’t want it in an ideal world, but because infinite append only data is scary. Keeping the data in the stash but not exposing it to users seems like the least desirable outcome, because users get no benefit but we pay the cost.

necaris commented:
> I saw that paragraph, but it seems to be about a similar issue in unrelated tables (source and sink status histories), and not about the problem of leaving the expired data on disk. 

I apologize for the confusion -- I was indeed referring to the same issue in those unrelated tables.

> Is it safe to delete things from the stash instead? Are we sure the billing code won't have a problem with that? This PR has the benefit that we could still go get the data if we really needed to.

Ideally at this point it should be!

> Maybe I'm missing something, but this PR seems to be lacking the key part where the expired storage usage records are actually removed from the stash. As far as I can tell, the expired records are simply not emitted into mz_storage_usage_by_shard table, but they remain in the stash.

My (mistaken) assumption had been that when `consolidate()` is called on the Stash the un-referenced rows would be compacted away. I will follow up with a draft to sort that out!

necaris commented:
See https://github.com/MaterializeInc/materialize/pull/17753 for the promised draft!
